### PR TITLE
Quantisation

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Resources:
- https://pytorch.org/blog/introduction-to-quantization-on-pytorch
- https://pytorch.org/docs/stable/quantization.html
- https://spell.ml/blog/pytorch-quantization-X8e7wBAAACIAHPhT
- https://leimao.github.io/blog/PyTorch-Quantization-Aware-Training/
- https://github.com/pytorch/fairseq/blob/master/examples/quant_noise/README.md

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/quantization.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+def wrap_quantize_forward_context(
+    quant_cb, model: pl.core.LightningModule, func: Callable, trigger_func: Optional[Callable] = None
+) -> Callable:
+    """
+    This decorator is used as context manager...
+    """
+
+    # todo: consider using registering hook before/after forward
+    def wrapper(data) -> Any:
+        _quent_run = trigger_func is None or trigger_func(model.trainer)
```
Comment (tchaton) [inline pytorch_lightning/callbacks/quantization.py:38]: typo. Should be quant.

File: pytorch_lightning/callbacks/quantization.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        else:
+            return False
+    else:
+        return state & hasattr(obj, attribs)
+
+
+class QuantizationAwareTraining(Callback):
+    """
+    Quantization allows speed-up inference and decrease model size by performing computations and storing tensors
+     at lower bitwidths than floating point precision such as INT8 or FLOAT16.
```
Comment (tchaton) [inline pytorch_lightning/callbacks/quantization.py:67]: Not sure `FLOAT16` makes sense, as people can train with AMP.

File: pytorch_lightning/callbacks/quantization.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    We use native PyTorch API so for more information see
+     `Quantization <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>_`
+    """
+
+    def __init__(
+        self,
+        qconfig: Union[str, QConfig] = 'fbgemm',
+        lambda_trigger: Optional[Callable] = None,
```
Comment (tchaton) [inline pytorch_lightning/callbacks/quantization.py:75]: Let's rename `collect_quantization` and let's make it bool or Callable.

Comment (tchaton) [inline pytorch_lightning/callbacks/quantization.py:75]: Let's add observer_type="moving_average"

File: pytorch_lightning/callbacks/quantization.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                to find what layer types can be fue=sed check https://github.com/pytorch/pytorch/pull/43286
+        """
+        if not isinstance(qconfig, (str, QConfig)):
+            raise MisconfigurationException(f"Unsupported qconfig: f{self._qconfig}")
+        self._qconfig = qconfig
+        self._lambda_trigger = lambda_trigger
+        self.modules_to_fuse = modules_to_fuse
+        self._forward_calls = 0
+
+    def increment_forward(self):
```
Comment (tchaton) [inline pytorch_lightning/callbacks/quantization.py:106]: Why this ?

File: pytorch_lightning/callbacks/quantization.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+        if self.modules_to_fuse:
+            self._check_feasible_fuse(pl_module)
+            torch.quantization.fuse_modules(pl_module, self.modules_to_fuse, inplace=True)
+
+        # Prepare the model for QAT. This inserts observers and fake_quants in
+        # the model that will observe weight and activation tensors during calibration.
+        torch.quantization.prepare_qat(pl_module, inplace=True)
+
+    def on_fit_end(self, trainer, pl_module):
```
Comment (tchaton) [inline pytorch_lightning/callbacks/quantization.py:145]: model should be cleaned out of extra modules and wrapper no ?

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1745,3 +1746,10 @@ def __get_hparams_assignment_variable(self):
             return "hparams"
 
         return None
+
+    def model_size(self) -> float:
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1750]: Let's make this a property

File: tests/callbacks/test_quantization.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    Trainer(callbacks=[qcb], **trainer_args).fit(qmodel, datamodule=dm)
+    quant_size = qmodel.model_size()
+    quant_mae = torch.mean(torch.tensor([model.measure(qmodel(x), y) for x, y in dm.test_dataloader()]))
+
+    # test that the trained model is smaller then initial
+    size_ratio = quant_size / org_size
+    assert size_ratio < 0.55
+    # test that the test score is almost the same as with pure training
+    diff_mae = abs(org_mae - quant_mae)
+    assert diff_mae < 15
```
Comment (tchaton) [inline tests/callbacks/test_quantization.py:120]: Can you use relative absolute instead.

File: tests/callbacks/test_quantization.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+    qmodel = LinearModel()
+    fusing_layers = [(f'layers.mlp_{i}', f'layers.mlp_{i}a') for i in range(3)]
+    qcb = QuantizationAwareTraining(modules_to_fuse=fusing_layers)
+    Trainer(callbacks=[qcb], **trainer_args).fit(qmodel, datamodule=dm)
+    quant_size = qmodel.model_size()
+    quant_mae = torch.mean(torch.tensor([model.measure(qmodel(x), y) for x, y in dm.test_dataloader()]))
+
+    # test that the trained model is smaller then initial
+    size_ratio = quant_size / org_size
```
Comment (tchaton) [inline tests/callbacks/test_quantization.py:116]: Same here, relative absolute.

**Review by tchaton — COMMENTED** [review]
Great work and amazing new feature !

File: pytorch_lightning/callbacks/quantization.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py]: droppped

File: pytorch_lightning/callbacks/quantization.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:181]: that is a good question, if you clean it you can pass to the model only quant inputs, if we keep these layer quant/dequant you can keep reading anything as before and still hold all other layers quantized -> small

Comment (tchaton) [reply pytorch_lightning/callbacks/quantization.py:181]: Let's keep it then. Can you still torch-script the model with the wrapper ?

Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:181]: I have cleaned it so let's see :]

Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:185]: I am curious if you could do it with hook even if you in fact insert an extra layer on-the-fly which is missing in the model init declaration

File: tests/callbacks/test_quantization.py (hunk 8 above)
Comment (Borda) [reply tests/callbacks/test_quantization.py:52]: done

File: tests/callbacks/test_quantization.py (hunk 7 above)
Comment (Borda) [reply tests/callbacks/test_quantization.py]: done

**Comment by Borda** [comment]
fyi, moving adding the new models to #5861

File: pytorch_lightning/callbacks/__init__.py [hunk 9]
```diff
@@ -35,4 +36,6 @@
     'ProgressBar',
     'ProgressBarBase',
     'ModelPruning',
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/__init__.py:38]: Could you make sure these are in alphabetical order (seems only `ModelPruning` is out of place)

File: pytorch_lightning/callbacks/quantization.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        attrib, attribs = attribs.split('.', 1)
+        if hasattr(obj, attrib):
+            return _recursive_hasattr(getattr(obj, attrib), attribs, state)
+        else:
+            return False
+    else:
+        return state & hasattr(obj, attribs)
+
+
+class QuantizationAwareTraining(Callback):
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/quantization.py:84]: Any particular reason it is not just called `Quantization`?

File: pytorch_lightning/callbacks/quantization.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+            modules_to_fuse: allows you fuse a few layer together as you can see in `diagram
+             <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>_`
+                to find what layer types can be fue=sed check https://github.com/pytorch/pytorch/pull/43286
+            input_compatible: preserve quent/dequant layers which allows to feat any input as to the original model,
+                but break compatibility to torchscript
+        """
+        if not isinstance(qconfig, (str, QConfig)):
+            raise MisconfigurationException(f"Unsupported qconfig: f{qconfig}.")
+        self._qconfig = qconfig
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/quantization.py:130]: check for correct string?
```suggestion
        if isinstance(qconfig, str) and qconfig not in ('fbgemm', 'qnnpack'):
            raise MisconfigurationException(f"Unsupport string for qconfig")
        self._qconfig = qconfig
```

File: pytorch_lightning/callbacks/quantization.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+                        return trainer.current_epoch == (trainer.max_epochs - 1)
+
+                    QuantizationAwareTraining(lambda_trigger=custom_trigger_last)
+
+            modules_to_fuse: allows you fuse a few layer together as you can see in `diagram
+             <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>_`
+                to find what layer types can be fue=sed check https://github.com/pytorch/pytorch/pull/43286
+            input_compatible: preserve quent/dequant layers which allows to feat any input as to the original model,
+                but break compatibility to torchscript
+        """
```
Comment (SkafteNicki) [inline pytorch_lightning/callbacks/quantization.py:127]: Add an example?

File: pytorch_lightning/metrics/functional/mean_relative_error.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    sum_rltv_error = torch.sum(torch.abs((preds - target) / target_nz))
+    n_obs = target.numel()
+    return sum_rltv_error, n_obs
+
+
+def _mean_relative_error_compute(sum_rltv_error: torch.Tensor, n_obs: int) -> torch.Tensor:
+    return sum_rltv_error / n_obs
+
+
+def mean_relative_error(preds: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/functional/mean_relative_error.py:34]: add import to this file:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/functional/__init__.py

File: pytorch_lightning/metrics/functional/mean_relative_error.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    Computes mean relative error
+
+    Args:
+        pred: estimated labels
+        target: ground truth labels
+
+    Return:
+        Tensor with MAE
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/functional/mean_relative_error.py:43]:
```suggestion
        Tensor with mean relative error
```

File: tests/callbacks/test_quantization.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+import torch
+
+from pytorch_lightning import seed_everything, Trainer
+from pytorch_lightning.callbacks import QuantizationAwareTraining
+from pytorch_lightning.metrics.functional.mean_relative_error import mean_relative_error
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests.helpers.zoo_data_models import RegressDataModule, RegressionModel
+
+
+@pytest.mark.parametrize("observe", ['histogram', 'average'])
```
Comment (SkafteNicki) [inline tests/callbacks/test_quantization.py:26]: do we test for different `qconfig` (string and custom qconfig)?

File: pytorch_lightning/metrics/functional/mean_relative_error.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    Example:
+
+        >>> x = torch.tensor([0., 1, 2, 3])
+        >>> y = torch.tensor([0., 1, 2, 2])
+        >>> mean_relative_error(x, y)
+        tensor(0.1250)
+
+    """
+    sum_rltv_error, n_obs = _mean_relative_error_update(preds, target)
+    return _mean_relative_error_compute(sum_rltv_error, n_obs)
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/functional/mean_relative_error.py:54]: class interface + test should be added, but I am fine with this being in a follow up PR :]

**Review by SkafteNicki — COMMENTED** [review]
Mix of comments :]

File: pytorch_lightning/callbacks/quantization.py (hunk 10 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:84]: there are a few other ways how to do quantization and this comes from official PT docs

File: Makefile [hunk 17]
```diff
@@ -3,7 +3,7 @@
 # to imitate SLURM set only single node
 export SLURM_LOCALID=0
 # assume you have installed need packages
-export SPHINX_MOCK_REQUIREMENTS=0
+export SPHINX_MOCK_REQUIREMENTS=1
```
Comment (Borda) [inline Makefile:6]:
```suggestion
export SPHINX_MOCK_REQUIREMENTS=0
```

File: pytorch_lightning/callbacks/quantization.py (hunk 11 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:130]: not sure if want to limit it here since we just pass the parameter to PT and if P adds an extra argument we would have to hardcode it here...

File: tests/callbacks/test_quantization.py (hunk 15 above)
Comment (Borda) [reply tests/callbacks/test_quantization.py:26]: we just test for the strings, the Custom is on user responsibility how he defines it...
in fact with passing string we get the default `qconfig` so yes we have it tested :]

File: pytorch_lightning/callbacks/quantization.py (hunk 10 above)
Comment (SeanNaren) [reply pytorch_lightning/callbacks/quantization.py:84]: Yeah keep it as is! the name is important

File: pytorch_lightning/callbacks/quantization.py (hunk 11 above)
Comment (SkafteNicki) [reply pytorch_lightning/callbacks/quantization.py:130]: alright, fine by me :]

File: pytorch_lightning/callbacks/quantization.py (hunk 10 above)
Comment (SkafteNicki) [reply pytorch_lightning/callbacks/quantization.py:84]: Okay, fine by me :]

File: pytorch_lightning/callbacks/quantization.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+                    def custom_trigger_last(trainer):
+                        return trainer.current_epoch == (trainer.max_epochs - 1)
+
+                    QuantizationAwareTraining(lambda_trigger=custom_trigger_last)
+
+            modules_to_fuse: allows you fuse a few layer together as you can see in `diagram
+             <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>_`
+                to find what layer types can be fue=sed check https://github.com/pytorch/pytorch/pull/43286
+            input_compatible: preserve quent/dequant layers which allows to feat any input as to the original model,
+                but break compatibility to torchscript
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:126]:
```suggestion
            input_compatible: preserve quant/dequant layers. This allows to feat any input as to the original model,
                but break compatibility to torchscript
```

Also, I don't understand what you mean with `This allows to feat any input as to the original model`

File: pytorch_lightning/callbacks/quantization.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+        self._observer_type = observer_type
+
+        if collect_quantization is not None and not isinstance(collect_quantization, (int, Callable)):
+            raise MisconfigurationException(
+                f'Unsupported `collect_quantization` "{observer_type}", allowed are int or callable.'
+            )
+        self._collect_quantization = collect_quantization
+
+        self.modules_to_fuse = modules_to_fuse
+        self._preserve_compatible = input_compatible
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:145]: why different names?

File: pytorch_lightning/callbacks/quantization.py [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+
+        if observer_type not in self.OBSERVER_TYPES:
+            raise MisconfigurationException(
+                f'Unsupported observer type "{observer_type}", allowed are {self.OBSERVER_TYPES}.'
+            )
+        self._observer_type = observer_type
+
+        if collect_quantization is not None and not isinstance(collect_quantization, (int, Callable)):
+            raise MisconfigurationException(
+                f'Unsupported `collect_quantization` "{observer_type}", allowed are int or callable.'
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:140]:
```suggestion
                f'Unsupported `collect_quantization` "{collect_quantization}", allowed are `int` or `Callable`.'
```

File: pytorch_lightning/callbacks/quantization.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+        if isinstance(self._qconfig, str):
+            if self._observer_type == 'histogram':
+                pl_module.qconfig = torch.quantization.get_default_qconfig(self._qconfig)
+            elif self._observer_type == 'average':
+                pl_module.qconfig = torch.quantization.get_default_qat_qconfig(self._qconfig)
+        elif isinstance(self._qconfig, QConfig):
+            pl_module.qconfig = self._qconfig
+
+        if self.modules_to_fuse:
+            self._check_feasible_fuse(pl_module)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:181]:
```suggestion
        if self._check_feasible_fuse(pl_module)
```

File: pytorch_lightning/core/memory.py [hunk 22]
```diff
@@ -227,6 +227,7 @@ def trainable_parameters(self) -> int:
 
     @property
     def model_size(self) -> float:
+        # todo: seems it does not work with quantized models
```
Comment (carmocca) [inline pytorch_lightning/core/memory.py:230]: question: what's the problem?

File: tests/callbacks/test_pruning.py [hunk 23]
```diff
@@ -20,15 +20,14 @@
 from torch import nn
 
 from pytorch_lightning import Trainer
+from pytorch_lightning.callbacks import ModelPruning
```
Comment (carmocca) [inline tests/callbacks/test_pruning.py:23]: Undo changes to this file? they are in the pruning PR already

File: pytorch_lightning/metrics/functional/mean_relative_error.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
+# Copyright The PyTorch Lightning team.
```
Comment (carmocca) [inline pytorch_lightning/metrics/functional/mean_relative_error.py:1]: Why not add this metric in a separate PR?

File: pytorch_lightning/callbacks/quantization.py [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+    """
+
+    # todo: consider using registering hook before/after forward
+    def wrapper(data) -> Any:
+        _is_func_true = isinstance(trigger_condition, Callable) and trigger_condition(model.trainer)
+        _is_count_true = isinstance(trigger_condition, int) and quant_cb._forward_calls < trigger_condition
+        _quant_run = trigger_condition is None or _is_func_true or _is_count_true
+        # apply custom trigger
+        if _quant_run:
+            quant_cb._forward_calls += 1
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:46]: Track quantization forward calls with the `trainer.dev_debugger`?

File: tests/callbacks/test_quantization.py [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+    dm = RegressDataModule()
+    qmodel = RegressionModel()
+    qcb = QuantizationAwareTraining(input_compatible=False)
+    Trainer(callbacks=[qcb], default_root_dir=tmpdir, max_epochs=1).fit(qmodel, datamodule=dm)
+
+    qmodel.to_torchscript()
+
+
+def test_quantization_exceptions(tmpdir):
+    """Test  wrong fuse layers"""
```
Comment (carmocca) [inline tests/callbacks/test_quantization.py:72]:
```suggestion
    """Test wrong fuse layers"""
```

File: tests/callbacks/test_quantization.py [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+    quant_size = qmodel.model_size
+    quant_score = torch.mean(torch.tensor([mean_relative_error(qmodel(x), y) for x, y in dm.test_dataloader()]))
+    assert quant_calls == qcb._forward_calls
+
+    # test that the trained model is smaller then initial
+    size_ratio = quant_size / org_size
+    assert size_ratio < 0.65
+    # test that the test score is almost the same as with pure training
+    diff_score = abs(org_score - quant_score)
+    assert diff_score < 0.35
```
Comment (carmocca) [inline tests/callbacks/test_quantization.py:58]: You can use `torch.testing.assert_allclose` and adjust the tolerance

https://github.com/pytorch/pytorch/blob/b9acfcddeb3b320f142f07d8051e5e344480f939/torch/testing/__init__.py#L209

File: pytorch_lightning/callbacks/quantization.py [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+        if self.modules_to_fuse:
+            self._check_feasible_fuse(pl_module)
+            torch.quantization.fuse_modules(pl_module, self.modules_to_fuse, inplace=True)
+
+        # Prepare the model for QAT. This inserts observers and fake_quants in
+        # the model that will observe weight and activation tensors during calibration.
+        torch.quantization.prepare_qat(pl_module, inplace=True)
+
+    def on_fit_end(self, trainer, pl_module):
+        pl_module.eval()
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:189]: why do we set eval here? Can't the user call `.eval()` after the model is converted?

File: tests/callbacks/test_quantization.py [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+from pytorch_lightning.callbacks import QuantizationAwareTraining
+from pytorch_lightning.metrics.functional.mean_relative_error import mean_relative_error
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests.helpers.datamodules import RegressDataModule
+from tests.helpers.simple_models import RegressionModel
+
+
+@pytest.mark.parametrize("observe", ['histogram', 'average'])
+@pytest.mark.parametrize("fuse", [True, False])
+def test_quantization(tmpdir, observe, fuse):
```
Comment (carmocca) [inline tests/callbacks/test_quantization.py:29]: Can we add gpu and ddp tests?

**Comment by carmocca** [comment]
So why do we only support QAT? Do we want to also provide static/dynamic quantization?

**Comment by Borda** [comment]
> So why do we only support QAT? Do we want to also provide static/dynamic quantization?

for static nor dynamic you do not need a training loop (for dynamic not even any data) so it is not big to integrate it in PL or what do you mean?

File: pytorch_lightning/callbacks/quantization.py (hunk 28 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:189]: from docs, it seems that you have to have a model in eval mode to perform conversion... (I may try it without)

File: pytorch_lightning/callbacks/quantization.py (hunk 25 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:46]: any special reason for it? we need it as a user argument anyway...

File: pytorch_lightning/metrics/functional/mean_relative_error.py (hunk 24 above)
Comment (Borda) [reply pytorch_lightning/metrics/functional/mean_relative_error.py:1]: no special reason

File: pytorch_lightning/callbacks/quantization.py [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+        data = func(data)
+        # apply custom trigger
+        if _quant_run:
+            data = model.dequant(data)
+        return data
+
+    return wrapper
+
+
+def wrap_quantize_forward_context(model: pl.core.LightningModule, func: Callable, quant, dequant) -> Callable:
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:57]: I want also drop the cb link and then the wrap func would be quite complex, more then needed

File: pytorch_lightning/core/memory.py (hunk 22 above)
Comment (Borda) [reply pytorch_lightning/core/memory.py:230]: if you have a large model and a slow filesystem it can take some time

File: pytorch_lightning/callbacks/quantization.py (hunk 25 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/quantization.py:46]: Isn't _foward_calls only used for the tests?

File: pytorch_lightning/core/memory.py (hunk 22 above)
Comment (carmocca) [reply pytorch_lightning/core/memory.py:230]: And saving it to a file is faster than that?

We could cache the model size

File: pytorch_lightning/callbacks/quantization.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/callbacks/quantization.py:57]: Sorry, I didn't get you. What `cb link`?

File: pytorch_lightning/callbacks/quantization.py (hunk 12 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:125]: well the example is just creating cb with all default atribs

File: pytorch_lightning/callbacks/quantization.py (hunk 25 above)
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:48]: also but the main reason is counter

File: pytorch_lightning/metrics/functional/mean_relative_error.py (hunk 24 above)
Comment (Borda) [reply pytorch_lightning/metrics/functional/mean_relative_error.py:1]: well I need it here to measure regression

Comment (carmocca) [reply pytorch_lightning/metrics/functional/mean_relative_error.py:1]: As mentioned below, I think you can use `torch.testing.assert_allclose` and adjust the tolerance

https://github.com/pytorch/pytorch/blob/b9acfcddeb3b320f142f07d8051e5e344480f939/torch/testing/__init__.py#L209

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/__init__.py (hunk 9 above)
Comment (Borda) [reply pytorch_lightning/callbacks/__init__.py:38]: probably in the next iteration and we shall do it in all `__all__`

File: pytorch_lightning/core/memory.py (hunk 22 above)
Comment (Borda) [reply pytorch_lightning/core/memory.py:230]: hehl sorry I swapped my answer with the model size above, this not sure if it was crashing or just returned the same size for quant and a standard model

Comment (Borda) [reply pytorch_lightning/core/memory.py]: ok, it returns 0.0 so added to the comment

**Comment by Borda** [comment]
@tchaton @SeanNaren @carmocca @SkafteNicki mind re-review 🐰

**Review by SeanNaren — APPROVED** [review]

**Comment by Borda** [comment]
@carmocca seems to be a very similar docs failer as yours...

**Comment by carmocca** [comment]
> @carmocca seems to be a very similar docs failer as yours...

Yes... Same failure

File: pytorch_lightning/callbacks/pruning.py [hunk 31]
```diff
@@ -27,10 +27,10 @@
 import pytorch_lightning as pl
 from pytorch_lightning.callbacks.base import Callback
 from pytorch_lightning.core.lightning import LightningModule
-from pytorch_lightning.utilities import _PYTORCH_PRUNE_AVAILABLE
+from pytorch_lightning.utilities import _TORCH_PRUNE_AVAILABLE
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 
-if _PYTORCH_PRUNE_AVAILABLE:
+if _TORCH_PRUNE_AVAILABLE:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:33]: Undoing these as it will generate conflicts with #5825 

```suggestion
from pytorch_lightning.utilities import _PYTORCH_PRUNE_AVAILABLE
from pytorch_lightning.utilities.exceptions import MisconfigurationException

if _PYTORCH_PRUNE_AVAILABLE:
```

File: tests/__init__.py [hunk 32]
```diff
@@ -31,3 +34,11 @@
 
 if not os.path.isdir(_TEMP_PATH):
     os.mkdir(_TEMP_PATH)
+
+_SKIPIF_ARGS_PT_LE_1_4 = dict(condition=_TORCH_LOWER_EQUAL_1_4, reason="test pytorch > 1.4")
```
Comment (carmocca) [inline tests/__init__.py:38]: Can you move all these unrelated test changes out of this PR? it will be easier to land

File: pytorch_lightning/callbacks/quantization.py [hunk 33]
```diff
@@ -0,0 +1,208 @@
...
+        qconfig: Union[str, QConfig] = 'fbgemm',
+        observer_type: str = "average",
+        collect_quantization: Optional[Union[int, Callable]] = None,
+        modules_to_fuse: Optional[Sequence] = None,
+        input_compatible: bool = True,
+    ) -> None:
+        """
+        Args:
+            qconfig: define quantization configuration see: `torch.quantization.QConfig
+             <https://pytorch.org/docs/stable/torch.quantization.html?highlight=qconfig#torch.quantization.QConfig>_`
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:113]:
```suggestion
             <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>_`
```

File: tests/__init__.py (hunk 32 above)
Comment (carmocca) [reply tests/__init__.py:38]: It's just a lot of changes which we want to properly do, so better do it separate and not pollute the quantization PR

Comment (Borda) [reply tests/__init__.py:38]: I have a feeling that when I told you similar on your PR you didn't listen... :P
and this is related to this addition as you can see that we need to filer tests for PT <= 1.4

**Review by SkafteNicki — APPROVED** [review]

File: tests/__init__.py (hunk 32 above)
Comment (carmocca) [reply tests/__init__.py:38]: Not the same because in my PR all changes were for pruning... Here there are changes to many other files

File: pytorch_lightning/metrics/functional/mean_relative_error.py (hunk 16 above)
Comment (Borda) [reply pytorch_lightning/metrics/functional/mean_relative_error.py:54]: next PR :]

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/callbacks/quantization.py [hunk 34]
```diff
@@ -0,0 +1,206 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Quantization
+^^^^^^^^^^^^
+
+.. warning::
+
+    Quantization is in beta and subject to change.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/quantization.py:20]:
```suggestion
```

I believe this is not necessary as it is not rendered anywhere (AFAIK)

**Review by carmocca — APPROVED** [review]
LGTM overall.

We can iterate on the _SKIPIF_ARGS thing in separate PR

Also, maybe we should create an issue for the model size TODO. Seems very hacky

File: pytorch_lightning/callbacks/quantization.py [hunk 35]
```diff
@@ -0,0 +1,206 @@
...
+
+        if self._check_feasible_fuse(pl_module):
+            torch.quantization.fuse_modules(pl_module, self.modules_to_fuse, inplace=True)
+
+        # Prepare the model for QAT. This inserts observers and fake_quants in
+        # the model that will observe weight and activation tensors during calibration.
+        torch.quantization.prepare_qat(pl_module, inplace=True)
+
+    def on_fit_end(self, trainer, pl_module):
+        # pl_module.eval()
```
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/callbacks/quantization.py:196]: just thinking that I would rather keep as it may happen that model without validation may end in a train stage

File: Makefile [hunk 36]
```diff
@@ -3,7 +3,7 @@
 # to imitate SLURM set only single node
 export SLURM_LOCALID=0
 # assume you have installed need packages
-export SPHINX_MOCK_REQUIREMENTS=1
+export SPHINX_MOCK_REQUIREMENTS=0
```
Comment (Borda) [inline Makefile:6]:
```suggestion
export SPHINX_MOCK_REQUIREMENTS=1
```
